pagopa/io-app
# chore: [IOAPPFD0-10] Startup saga, onboarding refactoring

# ⚠️ This pr depends on #4147 ⚠️
Please review #4147 first in order to avoid duplicate review work.

## Short description
This PR refactors the onboarding branching on the startup saga, in order to have a cleaner code

## List of changes proposed in this pull request
This PR aims to simplify the startup saga by re-arranging the onboarding branch condition.

The current state of the code is shown in the following diagram

![00](https://user-images.githubusercontent.com/5150343/203326001-5b3b01c9-f770-4d5a-a6e5-79d880e3d354.png)

The simplification process focuses on removing the direct flow on the right side of the diagram, which skips all the intermediate calls.

That condition happens when these three conditions are all met: there is a previous session token, there is a local stored pin and the session was refreshed. Specifically, we focus on the last of the three conditions (i.e., the session was refreshed).
If it is met, then the `previousSessionToken` must be different from the `sessionToken`. This happens only if the `previousSessionToken` value is `falsy` (otherwise, the two tokens will have the same value and the condition is not met). Now, having a `falsy` `previousSessionToken` means that the branch containing the condition is not even executed, since the first of the initial three conditions (i.e., there is a previous session token) is not met anymore.

With the consideration above, we deduct that the flow is never executed and can be safely removed, leading to the following simplified diagram

![01](https://user-images.githubusercontent.com/5150343/203326040-b17de036-1399-42c2-b7e3-4cdecd56c44e.png)

From this, we can see that there are some common steps, executed on both branches. Our proposal is to completely remove the branching and execute the steps using guards that maps the branching condition, formally:

- current execution has both session and pin;
- current execution has no session or no pin (which is the logical negation of the previous)

Other conditions are mapped adding additional guards to either one or the other.

This last diagram is the final code refactoring that is being proposed with this PR:

![02](https://user-images.githubusercontent.com/5150343/203326072-749fb487-8e20-4cb0-acdd-b8e1b3c14c34.png)

Note that we need to compute the `isFirstOnboarding` parameter before `call(askServicesPreferencesModeOptin(isFirstOnboarding: boolean))`.

Also note that, the flow condition `(has no session OR has no pin) AND isFirstOnboarding` can be further simplified to just check `isFirstOnboarding`, since if the user does not have a session, the code flow is interrupted long before reaching this point and the same goes for the not having a pin.

## How to test
In order to test this PR you have to go through the onboarding process so start an instance of the io-app-dev-server with the `firstOnboarding` set to `true` and make sure to reinstall the application (i.e., no previously saved local data on the device/simulator). In this case, you should see all the onboarding screens (namely, ToS, Mixpanel OptIn, Pin Configuration, Fingerprint (on device), email ACK, push notifications OptIn, service configuration and onboarding completion).

After this, you can test again, by restarting the io-app-dev-server with the `firstOnboarding` flag set to `false` (do not reinstall the application). In this case, you should only see the PIN screen and the push notification OptIn.

Also check all the logout/reset features during the PIN verification and during the onboarding.
